src/dart/parse.rs

```diff
@@ -0,0 +1,1739 @@
+        if let Some(token) = self.cur {
+            token.as_ident().map_or(false, |x| x == s)
+        } else {
+            false
+        }
+    }
+
+    fn bump_raw(&mut self) {
+        if let Some(token) = self.cur {
+            token.advance_src_pos(&mut self.line, &mut self.col);
+        }
+        self.cur = self.tokens.next();
+    }
+
+    fn bump(&mut self) {
+        loop {
+            self.bump_raw();
+            match self.cur {
+                Some(Token::WhiteSpace(_)) => {}
+                _ => return,
+            }
+        }
+    }
+
+    fn expect_punctuation(&mut self, c: char) -> ParseResult<()> {
+        if self.cur != Some(Token::Punctuation(c)) {
+            expected!(self, Punctuation(c));
+        }
+        self.bump();
+        Ok(())
+    }
+
+    fn eat_punctuation(&mut self, c: char) -> bool {
+        self.try(|p| p.expect_punctuation(c)).is_ok()
+    }
+
+    fn is_punctuation(&self, c: char) -> bool {
+        self.probe(|p| p.eat_punctuation(c))
+    }
+
+    fn expect_punctuation2(&mut self, c1: char, c2: char) -> ParseResult<()> {
+        if !self.is_punctuation(c1) {
+            expected!(self, Punctuation2(c1, c2));
+        }
+        self.bump_raw();
+        if !self.is_punctuation(c2) {
+            expected!(self, Punctuation2(c1, c2));
+        }
+        self.bump();
+        Ok(())
+    }
+
+    fn eat_punctuation2(&mut self, c1: char, c2: char) -> bool {
+        self.try(|p| p.expect_punctuation2(c1, c2)).is_ok()
+    }
+
+    fn is_punctuation2(&self, c1: char, c2: char) -> bool {
+        self.probe(|p| p.eat_punctuation2(c1, c2))
+    }
+
+    fn expect_bin_op(&mut self, op: BinOp) -> ParseResult<()> {
+        for op2 in BinOp::values() {
+            if op2 != op && op2.as_str().starts_with(op.as_str()) {
+                if self.is_bin_op(op2) {
+                    expected!(self, BinOp(op));
+                }
+            }
+        }
+        for c in op.as_str().chars() {
+            if !self.is_punctuation(c) {
+                expected!(self, BinOp(op));
+            }
+            self.bump_raw();
+        }
+        if let Some(Token::WhiteSpace(_)) = self.cur {
+            self.bump();
+        }
+        Ok(())
+    }
+
+    fn eat_bin_op(&mut self, op: BinOp) -> bool {
+        self.try(|p| p.expect_bin_op(op)).is_ok()
+    }
+
+    fn is_bin_op(&self, op: BinOp) -> bool {
+        self.probe(|p| p.eat_bin_op(op))
+    }
+
+    fn expect_keyword(&mut self, s: &'static str) -> ParseResult<()> {
+        if !self.is_keyword(s) {
+            expected!(self, Keyword(s));
+        }
+        self.bump();
+        Ok(())
+    }
+
+    fn eat_keyword(&mut self, s: &'static str) -> bool {
+        self.expect_keyword(s).is_ok()
+    }
+
+    fn probe<F: FnOnce(&mut Self) -> R, R>(&self, f: F) -> R {
+        f(&mut self.clone())
+    }
+
+    fn try<F: FnOnce(&mut Self) -> ParseResult<T>, T>(&mut self, f: F) -> ParseResult<T> {
```

Comment:
This could probably return `Option` for it to be less misusable (thanks @anyska!).